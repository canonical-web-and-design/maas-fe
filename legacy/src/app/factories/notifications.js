/* Copyright 2017 Canonical Ltd.  This software is licensed under the
 * GNU Affero General Public License version 3 (see the file LICENSE).
 *
 * MAAS Notifications Manager.
 *
 * Manages notifications in the browser. Uses RegionConnection to load
 * notifications, await new, updated, and deleted notifications, and to
 * dismiss them.
 */
import angular from "angular";

function NotificationsManager(RegionConnection, Manager) {
  function NotificationsManager() {
    Manager.call(this);

    this._pk = "id";
    this._handler = "notification";

    // Listen for notify events for the notification object.
    RegionConnection.registerNotifier(
      "notification",
      angular.bind(this, this.onNotify)
    );
  }

  NotificationsManager.prototype = new Manager();
  NotificationsManager.prototype.dismiss = function (notification) {
    return RegionConnection.callMethod("notification.dismiss", {
      id: notification.id,
    });
  };
  NotificationsManager.prototype.getItems = function () {
    return this._items.filter(({ ident }) => ident !== "release_notification");
  };

  return new NotificationsManager();
}

NotificationsManager.$inject = ["RegionConnection", "Manager"];

export default NotificationsManager;
