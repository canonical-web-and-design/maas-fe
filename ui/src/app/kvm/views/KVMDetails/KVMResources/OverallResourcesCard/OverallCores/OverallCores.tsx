import Meter from "app/base/components/Meter";
import { COLOURS } from "app/base/constants";
import type { Pod, PodResource } from "app/store/pod/types";
import { resourceWithOverCommit } from "app/store/pod/utils";

type Props = {
  cores: PodResource;
  overCommit: Pod["cpu_over_commit_ratio"];
};

const OverallCores = ({ cores, overCommit }: Props): JSX.Element => {
  const { allocated_other, allocated_tracked, free } = resourceWithOverCommit(
    cores,
    overCommit
  );
  const total = allocated_other + allocated_tracked + free;
  return (
    <div className="overall-cores">
      <h4 className="p-heading--small u-sv1">CPU cores</h4>
      <div className="overall-cores__meter">
        <div className="u-align--right">
          <p className="p-heading--small u-text--light">
            Project
            <span className="u-nudge-right--small">
              <i className="p-circle--positive u-no-margin--top"></i>
            </span>
          </p>
          <div className="u-nudge-left">{allocated_tracked}</div>
        </div>
        <div className="u-align--right">
          <p className="p-heading--small u-text--light">
            Others
            <span className="u-nudge-right--small">
              <i className="p-circle--link u-no-margin--top"></i>
            </span>
          </p>
          <div className="u-nudge-left">{allocated_other}</div>
        </div>
        <div className="u-align--right">
          <p className="p-heading--small u-text--light">
            Free
            <span className="u-nudge-right--small">
              <i className="p-circle--link-faded u-no-margin--top"></i>
            </span>
          </p>
          <div className="u-nudge-left" data-test="free">
            {free}
          </div>
        </div>
        <div style={{ gridArea: "meter" }}>
          <Meter
            data={[
              {
                color: COLOURS.POSITIVE,
                value: allocated_tracked,
              },
              {
                color: COLOURS.LINK,
                value: allocated_other,
              },
              {
                color: COLOURS.LINK_FADED,
                // A negative free value implies cores have been over-committed,
                // but we don't actually want to represent negative values in
                // the chart.
                value: free > 0 ? free : 0,
              },
            ]}
            max={total}
            segmented
            small
          />
        </div>
      </div>
      <div className="u-align--right">
        <h4 className="p-heading--small u-no-max-width u-text--muted">Total</h4>
        <div>{total}</div>
      </div>
    </div>
  );
};

export default OverallCores;
