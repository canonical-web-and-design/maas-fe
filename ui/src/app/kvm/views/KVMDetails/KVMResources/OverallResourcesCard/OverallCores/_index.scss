@mixin OverallCores {
  .overall-cores__meter {
    display: grid;
    grid-column-gap: $sph-inner;
    grid-template-areas:
      "project others free"
      "meter meter meter";
    grid-template-columns: max-content 1fr 1fr;
    grid-template-rows: min-content;

    .p-meter--small {
      margin-bottom: 0;
    }
  }
}
