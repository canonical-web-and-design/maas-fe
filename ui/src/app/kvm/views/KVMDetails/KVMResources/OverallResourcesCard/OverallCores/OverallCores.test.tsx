import { mount, shallow } from "enzyme";

import OverallCores from "./OverallCores";

import { podResource as podResourceFactory } from "testing/factories";

describe("OverallCores", () => {
  it("renders", () => {
    const cores = podResourceFactory({
      allocated_tracked: 1,
      allocated_other: 2,
      free: 3,
    });
    const wrapper = shallow(<OverallCores cores={cores} overCommit={2} />);

    expect(wrapper).toMatchSnapshot();
  });

  it("can show whether cores have been over-committed", () => {
    const cores = podResourceFactory({
      allocated_tracked: 1,
      allocated_other: 2,
      free: -1,
    });
    const wrapper = mount(<OverallCores cores={cores} overCommit={1} />);

    expect(wrapper.find("[data-test='meter-overflow']").exists()).toBe(true);
  });
});
