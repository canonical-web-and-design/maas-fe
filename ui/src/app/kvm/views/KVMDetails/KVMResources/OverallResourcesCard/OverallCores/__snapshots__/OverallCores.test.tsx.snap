// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OverallCores renders 1`] = `
<div
  className="overall-cores"
>
  <h4
    className="p-heading--small u-sv1"
  >
    CPU cores
  </h4>
  <div
    className="overall-cores__meter"
  >
    <div
      className="u-align--right"
    >
      <p
        className="p-heading--small u-text--light"
      >
        Project
        <span
          className="u-nudge-right--small"
        >
          <i
            className="p-circle--positive u-no-margin--top"
          />
        </span>
      </p>
      <div
        className="u-nudge-left"
      >
        1
      </div>
    </div>
    <div
      className="u-align--right"
    >
      <p
        className="p-heading--small u-text--light"
      >
        Others
        <span
          className="u-nudge-right--small"
        >
          <i
            className="p-circle--link u-no-margin--top"
          />
        </span>
      </p>
      <div
        className="u-nudge-left"
      >
        2
      </div>
    </div>
    <div
      className="u-align--right"
    >
      <p
        className="p-heading--small u-text--light"
      >
        Free
        <span
          className="u-nudge-right--small"
        >
          <i
            className="p-circle--link-faded u-no-margin--top"
          />
        </span>
      </p>
      <div
        className="u-nudge-left"
        data-test="free"
      >
        9
      </div>
    </div>
    <div
      style={
        Object {
          "gridArea": "meter",
        }
      }
    >
      <Meter
        data={
          Array [
            Object {
              "color": "#0E8420",
              "value": 1,
            },
            Object {
              "color": "#0066CC",
              "value": 2,
            },
            Object {
              "color": "#D3E4ED",
              "value": 9,
            },
          ]
        }
        max={12}
        segmented={true}
        small={true}
      />
    </div>
  </div>
  <div
    className="u-align--right"
  >
    <h4
      className="p-heading--small u-no-max-width u-text--muted"
    >
      Total
    </h4>
    <div>
      12
    </div>
  </div>
</div>
`;
